packages/eslint-plugin/src/rules/prefer-at.ts

```diff
@@ -0,0 +1,219 @@
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import * as path from 'path';
+import * as ts from 'typescript';
+
+import * as util from '../util';
+
+export default util.createRule({
+  name: 'prefer-at',
+  meta: {
+    type: 'suggestion',
+    fixable: 'code',
+    docs: {
+      description:
+        'Enforce the use of `array.at(-1)` instead of `array[array.length - 1]`',
+      recommended: false,
+      requiresTypeChecking: true,
+    },
+    messages: {
+      preferAt:
+        'Expected a `{{name}}.at(-1)` instead of `{{name}}[{{name}}.length - 1]`.',
+    },
+    schema: [
+      {
+        oneOf: [
+          {
+            type: 'object',
+            properties: {
+              ignoreFunctions: {
+                type: 'boolean',
+              },
+            },
+            additionalProperties: false,
+            required: ['ignoreFunctions'],
+          },
+        ],
+      },
+    ],
+  },
+  defaultOptions: [
+    {
+      ignoreFunctions: false,
+    },
+  ],
+  create(context, [options]) {
+    const parserServices = util.getParserServices(context);
+    const checker = parserServices.program.getTypeChecker();
+
+    class UnknownNodeError extends Error {
+      public constructor(node: TSESTree.Node) {
+        super(`UnknownNode ${node.type}`);
+      }
+    }
+
+    function getName(node: TSESTree.Node): string {
+      switch (node.type) {
+        case AST_NODE_TYPES.Identifier:
+          return node.name;
+        case AST_NODE_TYPES.MemberExpression:
+          return getName(node.property);
+        default:
+          throw new UnknownNodeError(node);
+      }
+    }
+
+    function getFullName(node: TSESTree.Node): string {
+      switch (node.type) {
+        case AST_NODE_TYPES.PrivateIdentifier:
+          return `#${node.name}`;
+        case AST_NODE_TYPES.Identifier:
+          return node.name;
+        case AST_NODE_TYPES.Literal:
+          return node.raw;
+        case AST_NODE_TYPES.ThisExpression:
+          return 'this';
+        case AST_NODE_TYPES.CallExpression:
+          return `${getFullName(node.callee)}()`;
+        case AST_NODE_TYPES.MemberExpression:
+          if (node.property.type === AST_NODE_TYPES.Literal) {
+            return `${getFullName(node.object)}[${node.property.raw}]`;
+          }
+          return `${getFullName(node.object)}.${getFullName(node.property)}`;
+        default:
+          throw new UnknownNodeError(node);
+      }
+    }
+
+    function hasCallExpression(node: TSESTree.MemberExpression): boolean {
+      return (
+        node.object.type === AST_NODE_TYPES.CallExpression ||
+        (node.object.type === AST_NODE_TYPES.MemberExpression &&
+          hasCallExpression(node.object))
+      );
+    }
+
+    function getTypeAtLocation(node: TSESTree.Node): ts.Type | undefined {
+      const originalNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+      if (!originalNode) {
+        return undefined;
```

Comment:
[Request] Indeed, the type system says `originalNode` is never `undefined`, and this line is never hit by tests. It can be removed. This function returns `ts.Type`.